File: CP/LeetCode/C/27.c
// 27. Remove Element

#include <stdio.h>

int removeElement(int *nums, int numSize, int val)
{
    int index = 0;

    for (int i = 0; i < numSize; i++)
    {
        if (nums[i] != val)
        {
            nums[index] = nums[i];
            index++;
        }
    }

    return index;
}

int main()
{
    int nums[] = {3, 2, 2, 3};
    int numSize = sizeof(nums) / sizeof(nums[0]);
    int val = 3;

    int newLength = removeElement(nums, numSize, val);

    printf("New length: %d\n", newLength);
    printf("Modified array: ");
    for (int i = 0; i < newLength; i++)
    {
        printf("%d ", nums[i]);
    }
    printf("\n");

    return 0;
}

File: CP/LeetCode/Python/27.py
# 27. Remove Element


class Solution:
    def removeElement(self, nums: list[int], val: int) -> int:
        # while val in nums:
        #     nums.remove(val)

        # return len(nums)

        if not nums:
            return 0

        write_index = 0

        for read_index in range(len(nums)):
            if nums[read_index] != val:
                nums[write_index] = nums[read_index]
                write_index += 1

        return write_index


if __name__ == "__main__":
    nums = [3, 2, 2, 3]
    val = 3

    solution = Solution()
    new_length = solution.removeElement(nums, val)

    print(f"New length: {new_length}")
    print(f"Modified array: {nums[:new_length]}")
